### PR TITLE
allow use of byte_buf_reserve on buffers with initial capacity 0

### PR DESCRIPTION
**Pull request by ilevyor**
*Issue #, if available:*

*Description of changes:*
edited byte-buf-reserve() to work with byte_buffers whose size is initially 0

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/byte_buf.c [hunk 1]
```diff
@@ -750,7 +750,9 @@ int aws_byte_buf_reserve(struct aws_byte_buf *buffer, size_t requested_capacity)
         AWS_POSTCONDITION(aws_byte_buf_is_valid(buffer));
         return AWS_OP_SUCCESS;
     }
-
+    if (!buffer->buffer && !buffer->capacity && requested_capacity > buffer->capacity) {
+        return aws_byte_buf_init(buffer, buffer->allocator, requested_capacity);
```
Comment (bretambrose) [inline source/byte_buf.c:754]: Probably needs an error check and AWS_POSTCONDITION here

File: tests/byte_buf_test.c [hunk 2]
```diff
@@ -761,6 +761,17 @@ static int s_test_byte_buf_reserve(struct aws_allocator *allocator, void *ctx) {
 }
 AWS_TEST_CASE(test_byte_buf_reserve, s_test_byte_buf_reserve)
 
+static int s_test_byte_buf_reserve_initial_capacity_zero(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+    struct aws_byte_buf buffer;
+    AWS_ZERO_STRUCT(buffer);
+    ASSERT_SUCCESS(aws_byte_buf_init(&buffer, allocator, 0));
+    ASSERT_SUCCESS(aws_byte_buf_reserve(&buffer, 1));
+    aws_byte_buf_clean_up(&buffer);
```
Comment (bretambrose) [inline tests/byte_buf_test.c:770]: I'd verify the capacity too here.

File: source/byte_buf.c (hunk 1 above)
Comment (ilevyor) [reply source/byte_buf.c:754]: Why isn't the post condition in aws_byte_buf_init sufficient?

Comment (bretambrose) [reply source/byte_buf.c]: Not knowing much about the proof system, I don't know if it is or isn't.

**Review by bretambrose — APPROVED** [review]

**Pull request merged**